src/Datadog.Trace/Tagging/TagsList.cs

```diff
@@ -252,48 +259,42 @@ private int WriteTags(ref byte[] bytes, int offset)
 
             int count = 0;
 
-            var tags = Tags;
-            var additionalTags = GetAdditionalTags();
+            // We don't know the final count yet, write a fixed-size header and note the offset
+            var countOffset = offset;
+            offset += MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, offset, 0);
 
-            foreach (var property in additionalTags)
-            {
-                if (property.Getter(this) != null)
-                {
-                    count++;
-                }
-            }
+            var tags = Tags;
 
             if (tags != null)
             {
                 lock (tags)
                 {
                     count += tags.Count;
 
-                    offset += MessagePackBinary.WriteMapHeader(ref bytes, offset, count);
-
                     foreach (var pair in tags)
                     {
-                        offset += MessagePackBinary.WriteString(ref bytes, offset, pair.Key);
-                        offset += MessagePackBinary.WriteString(ref bytes, offset, pair.Value);
+                        WriteTag(ref bytes, ref offset, pair.Key, pair.Value);
```

Comment:
Shouldn't we check these tag values for null like we do for the additional tags below?
We never want to send null tag values to the Agent because it breaks deserialization. (Maybe this is fixed in newer versions, but we should avoid it if possible.)

Same comment for metrics below.

Reply:
I think the `SetTag` implementation ensures it's not null as far as I can see? 🤔

Reply:
Yes, but these well-known tags are set via properties and not stored in a dictionary, like the "additional tags" set using `SetTag`. For example you can set `WebTags.HttpMethod = null`
https://github.com/DataDog/dd-trace-dotnet/blob/5bfda09d1670c84846371fba35eb4d546949fc54/src/Datadog.Trace/Tagging/WebTags.cs#L17

Reply:
You're right. I misread `SetTag()`. For some reason I thought `GetAdditionalTags()` returned the tags from the list, not the properties. I had them reversed.

Reply:
As mentioned by Andrew, we null-check the tags bound to properties because we can't prevent them from being null, but for the others we catch that in `SetTag`